### PR TITLE
External ca api

### PR DESCRIPTION
**Pull request by EItanya**

### CODE REVIEW

File: api/certificates/v1/issued_certificate.proto [hunk 1]
```diff
@@ -37,13 +38,15 @@ message IssuedCertificateSpec {
...
-    // The organization for this certificate.
+    // Deprecated in favor of `common_cert_options.org_name`
     string org = 2;
 
+    // Deprecated in favor of `gloo_mesh_ca.signing_certificate_secret`
     // The secret containing the root SSL certificate used to sign this IssuedCertificate (located in the certificate issuer's cluster).
     .core.skv2.solo.io.ObjectRef signing_certificate_secret = 3;
 
     // The secret containing the SSL certificate to be generated for this IssuedCertificate (located in the Gloo Mesh agent's cluster).
+    // If this value is nil, it means it is the responsibility of the sidecar agent to write the data to memory (Enterprise only)
```
Comment (harveyxia) [inline api/certificates/v1/issued_certificate.proto:49]:
```suggestion
    // If nil, the sidecar agent stores the signing certificate in memory. (Enterprise only)
```

File: api/certificates/v1/issued_certificate.proto [hunk 2]
```diff
@@ -37,13 +38,15 @@ message IssuedCertificateSpec {
      */
     repeated string hosts = 1;
 
-    // The organization for this certificate.
+    // Deprecated in favor of `common_cert_options.org_name`
```
Comment (harveyxia) [inline api/certificates/v1/issued_certificate.proto:41]:
```suggestion
    // DEPRECATED: in favor of `common_cert_options.org_name`
```

Let's use all caps followed colon for marking deprecated, it's more visually apparent for the human reader

File: api/certificates/v1/issued_certificate.proto [hunk 3]
```diff
@@ -37,13 +38,15 @@ message IssuedCertificateSpec {
...
     string org = 2;
 
+    // Deprecated in favor of `gloo_mesh_ca.signing_certificate_secret`
```
Comment (harveyxia) [inline api/certificates/v1/issued_certificate.proto:44]:
```suggestion
    // DEPRECATED: in favor of `gloo_mesh_ca.signing_certificate_secret`
```

File: api/certificates/v1/issued_certificate.proto [hunk 4]
```diff
@@ -56,6 +59,36 @@ message IssuedCertificateSpec {
...
+        // Gloo Mesh CA options
+        GlooMeshCA gloo_mesh_ca = 7;        
+        // Agent CA options
+        AgentCA agent_ca = 8;
+    }
+
+}
+
+// Set of options which represent the certificate authorities the management cluster can use
+// to sign the intermediate certs
```
Comment (harveyxia) [inline api/certificates/v1/issued_certificate.proto:76]:
```suggestion
// to sign the intermediate certs.
```

File: api/certificates/v1/issued_certificate.proto [hunk 5]
```diff
@@ -56,6 +59,36 @@ message IssuedCertificateSpec {
...
+message GlooMeshCA {
+    // Signing method which gloo-mesh management will use to sign the intermediate cert
```
Comment (harveyxia) [inline api/certificates/v1/issued_certificate.proto:78]: Would this more accurately be described as the location of the signing secret? Same for below

File: api/common/v1/certificates.proto [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  string csr_path = 2;
+
+  // Server is the connection address for the Vault server, e.g: "https://vault.example.com:8200".
+  string server = 3;
+
+	// PEM encoded CA bundle used to validate Vault server certificate. Only used
+	// if the Server URL is using HTTPS protocol. This parameter is ignored for
+	// plain HTTP protocol connection. If not set the system root certificates
+	// are used to validate the TLS connection.
+  bytes CaBundle = 4;
```
Comment (harveyxia) [inline api/common/v1/certificates.proto:28]:
```suggestion
  bytes ca_bundle = 4;
```

File: api/certificates/v1/issued_certificate.proto (hunk 5 above)
Comment (EItanya) [reply api/certificates/v1/issued_certificate.proto:78]: Not necesserily, as there may not be any secret at all, that's why I described it as a signing method. That method could be a `Secret` or an external CA. Definitely open to other terms/verbiage here though

Comment (EItanya) [reply api/certificates/v1/issued_certificate.proto:78]: Just changed it Certificate authority, since that's what it really is

Comment (harveyxia) [reply api/certificates/v1/issued_certificate.proto]: I meant to say location of the signing CA 👍

File: api/certificates/v1/vault_ca.proto [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+option go_package = "github.com/solo-io/gloo-mesh/pkg/api/certificates.mesh.gloo.solo.io/v1";
+
+import "github.com/solo-io/skv2/api/core/v1/core.proto";
+
+import "extproto/ext.proto";
+option (extproto.equal_all) = true;
+
+message VaultCA {
+
+	// ca_path is the mount path of the Vault PKI backend's `sign` endpoint, e.g:
```
Comment (harveyxia) [inline api/certificates/v1/vault_ca.proto:12]: nit: fix the indentation on the docstrings in this file

Comment (harveyxia) [inline api/certificates/v1/vault_ca.proto:12]:
```suggestion
	// `ca_path` is the mount path of the Vault PKI backend's `sign` endpoint, e.g:
```

File: api/certificates/v1/vault_ca.proto [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+	// "my_pki_mount/sign/my-role-name".
+  string ca_path = 1;
+
+	// ca_path is the mount path of the Vault PKI backend's `generate` endpoint, e.g:
```
Comment (harveyxia) [inline api/certificates/v1/vault_ca.proto:16]:
```suggestion
	// `csr_path` is the mount path of the Vault PKI backend's `generate` endpoint, e.g:
```

File: api/certificates/v1/vault_ca.proto [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+	// "my_pki_mount/intermediate/generate/exported".
+  // exported is necessary here as istio needs access to the private key
```
Comment (harveyxia) [inline api/certificates/v1/vault_ca.proto:18]: is this line still relevant?

File: api/certificates/v1/vault_ca.proto [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  string server = 3;
+
+	// PEM encoded CA bundle used to validate Vault server certificate. Only used
+	// if the Server URL is using HTTPS protocol. This parameter is ignored for
+	// plain HTTP protocol connection. If not set the system root certificates
+	// are used to validate the TLS connection.
+  bytes CaBundle = 4;
+
+	// Name of the vault namespace. Namespaces is a set of features within Vault Enterprise that allows Vault environments to support Secure Multi-tenancy. e.g: "ns1"
+	// More about namespaces can be found here https://www.vaultproject.io/docs/enterprise/namespaces
```
Comment (harveyxia) [inline api/certificates/v1/vault_ca.proto:31]:
```suggestion
	// More about namespaces can be found [here](https://www.vaultproject.io/docs/enterprise/namespaces)
```

File: api/certificates/v1/vault_ca.proto [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+  string csr_path = 2;
+
+  // Server is the connection address for the Vault server, e.g: "https://vault.example.com:8200".
+  string server = 3;
+
+	// PEM encoded CA bundle used to validate Vault server certificate. Only used
+	// if the Server URL is using HTTPS protocol. This parameter is ignored for
+	// plain HTTP protocol connection. If not set the system root certificates
+	// are used to validate the TLS connection.
+  bytes CaBundle = 4;
```
Comment (harveyxia) [inline api/certificates/v1/vault_ca.proto:28]:
```suggestion
  bytes ca_bundle = 4;
```

File: api/certificates/v1/vault_ca.proto [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+}
+
+message VaultKubernetesAuth {
+  // The Vault mountPath here is the mount path to use when authenticating with
+  // Vault. For example, setting a value to `/v1/auth/foo`, will use the path
+  // `/v1/auth/foo/login` to authenticate with Vault. If unspecified, the
+  // default value "/v1/auth/kubernetes" will be used.
+  string path = 1;
```
Comment (harveyxia) [inline api/certificates/v1/vault_ca.proto:53]: Should this be named `mount_path`? If not, then let's have the docstring refer to `path`

File: api/certificates/v1/vault_ca.proto (hunk 9 above)
Comment (EItanya) [reply api/certificates/v1/vault_ca.proto:18]: Yes, I'll link to the vault doc explaining why I added that

File: api/certificates/v1/issued_certificate.proto [hunk 13]
```diff
@@ -37,13 +38,15 @@ message IssuedCertificateSpec {
...
-    // The organization for this certificate.
+    // DEPRECATED: in favor of `common_cert_options.org_name`
     string org = 2;
 
+    // DEPRECATED: in favor of `gloo_mesh_ca.signing_certificate_secret`
     // The secret containing the root SSL certificate used to sign this IssuedCertificate (located in the certificate issuer's cluster).
     .core.skv2.solo.io.ObjectRef signing_certificate_secret = 3;
 
     // The secret containing the SSL certificate to be generated for this IssuedCertificate (located in the Gloo Mesh agent's cluster).
+    // If nil, the sidecar agent stores the signing certificate in memory. (Enterprise only)
```
Comment (ilackarms) [inline api/certificates/v1/issued_certificate.proto:49]: i think we need to come with a more specific name for "sidecar agent" for use in our docs. Thinking about what happens if we add more agent sidecars to our system

Comment (ilackarms) [reply api/certificates/v1/issued_certificate.proto:49]: though we don't need to decide this now*

Comment (EItanya) [reply api/certificates/v1/issued_certificate.proto:49]: Agreed

**Review by ilackarms — APPROVED** [review]

**Pull request merged**